src/openms_gui/source/VISUAL/PlotCanvas.cpp

```diff
@@ -383,75 +388,69 @@ namespace OpenMS
+  }
 
+  bool PlotCanvas::addLayer(ExperimentSharedPtrType map, ODExperimentSharedPtrType od_map, const String & filename)
+  {
     // both empty
-    if (!new_layer.getPeakData()->getChromatograms().empty() 
-     && !new_layer.getPeakData()->empty())
+    if (!map->getChromatograms().empty() 
+     && !map->empty())
     {
       // TODO : handle this case better
       OPENMS_LOG_WARN << "Your input data contains chromatograms and spectra, falling back to display spectra only." << std::endl;
     }
 
+    LayerDataBaseUPtr new_layer;
     // check which one is empty
-    if (!new_layer.getPeakData()->getChromatograms().empty() 
-      && new_layer.getPeakData()->empty())
+    if (!map->getChromatograms().empty() 
+      && map->empty())
     {
-      new_layer.type = LayerData::DT_CHROMATOGRAM;
+      new_layer.reset(new LayerDataChrom);
     }
     else
     {
-      new_layer.type = LayerData::DT_PEAK;
+      new_layer.reset(new LayerDataPeak);
     }
+    new_layer->setPeakData(map);
+    new_layer->setOnDiscPeakData(od_map);
     
+    augmentLayer(new_layer.get(), param_, filename);
     layers_.addLayer(std::move(new_layer));
     return finishAdding_();
   }
 
   bool PlotCanvas::addLayer(FeatureMapSharedPtrType map, const String & filename)
   {
-    LayerData new_layer;
-    new_layer.param = param_;
-    new_layer.filename = filename;
-    new_layer.setName(QFileInfo(filename.toQString()).completeBaseName());
-    new_layer.getFeatureMap() = map;
-    new_layer.type = LayerData::DT_FEATURE;
+    LayerDataBaseUPtr new_layer(new LayerDataFeature);
+    new_layer->getFeatureMap() = map;
 
+    augmentLayer(new_layer.get(), param_, filename);
     layers_.addLayer(std::move(new_layer));
     return finishAdding_();
   }
 
   bool PlotCanvas::addLayer(ConsensusMapSharedPtrType map, const String & filename)
   {
-    LayerData new_layer;
-    new_layer.param = param_;
-    new_layer.filename = filename;
-    new_layer.setName(QFileInfo(filename.toQString()).completeBaseName());
-    new_layer.getConsensusMap() = map;
-    new_layer.type = LayerData::DT_CONSENSUS;
+    LayerDataBaseUPtr new_layer(new LayerDataConsensus);
+    new_layer->getConsensusMap() = map;
 
+    augmentLayer(new_layer.get(), param_, filename);
     layers_.addLayer(std::move(new_layer));
     return finishAdding_();
   }
 
-  bool PlotCanvas::addLayer(vector<PeptideIdentification> & peptides,
-                                const String & filename)
+  bool PlotCanvas::addLayer(vector<PeptideIdentification>& peptides, const String& filename)
   {
-    LayerData new_layer;
-    new_layer.param = param_;
-    new_layer.filename = filename;
-    new_layer.setName(QFileInfo(filename.toQString()).completeBaseName());
-    new_layer.peptides.swap(peptides);
-    new_layer.type = LayerData::DT_IDENT;
-
-    layers_.addLayer(std::move(new_layer));
+    LayerDataIdent* new_layer(new LayerDataIdent);
+    new_layer->getPeptideIds().swap(peptides);
```

Comment:
std::move ?

Reply:
or already an interface? setPeptideIds(...) ? could make transition to the new ID datastructure easier (if we store it internally in IDData to save memory)

Reply:
good point. Will extend the interface.